# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: extensions/permissions/PermissionDelegateHandler.cpp
# Commit: a2bec965260d
# Full Hash: a2bec965260d58e8f5bd814c179444875d9962cf
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1587743
# File Overlap Count: 1
# Description:
#   Bug 1587743 - Part 2: Removing mTopLevelPrincipal from the PermissionDelegateHandler. r=baku
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D79133
# ==============================================================================

diff -r 44e715808901 -r a2bec965260d extensions/permissions/PermissionDelegateHandler.cpp
--- a/extensions/permissions/PermissionDelegateHandler.cpp	Fri Jun 12 16:31:49 2020 +0000
+++ b/extensions/permissions/PermissionDelegateHandler.cpp	Fri Jun 12 16:31:56 2020 +0000
@@ -148,13 +148,29 @@
   }
 
   mPrincipal = mDocument->NodePrincipal();
-  nsPIDOMWindowInner* window = mDocument->GetInnerWindow();
-  nsGlobalWindowInner* innerWindow = nsGlobalWindowInner::Cast(window);
-  if (innerWindow) {
-    mTopLevelPrincipal = innerWindow->GetTopLevelAntiTrackingPrincipal();
+  return true;
+}
+
+static bool IsCrossOriginContentToTop(Document* aDocument) {
+  MOZ_ASSERT(aDocument);
+
+  BrowsingContext* topBrowsingContext = aDocument->GetBrowsingContext()->Top();
+
+  // In Fission, we can know if it is cross-origin by checking whether both
+  // contexts in the same process. So, If they are not in the same process, we
+  // can say that it's cross-origin.
+  if (!topBrowsingContext->IsInProcess()) {
+    return true;
   }
 
-  return true;
+  RefPtr<Document> topDoc = topBrowsingContext->GetDocument();
+  if (!topDoc) {
+    return true;
+  }
+
+  nsCOMPtr<nsIPrincipal> topLevelPrincipal = topDoc->NodePrincipal();
+
+  return !aDocument->NodePrincipal()->Subsumes(topLevelPrincipal);
 }
 
 bool PermissionDelegateHandler::HasFeaturePolicyAllowed(
@@ -189,7 +205,7 @@
 
   if (info->mPolicy == DelegatePolicy::ePersistDeniedCrossOrigin &&
       !mDocument->IsTopLevelContentDocument() &&
-      !mPrincipal->Subsumes(mTopLevelPrincipal)) {
+      IsCrossOriginContentToTop(mDocument)) {
     return false;
   }
 
@@ -225,17 +241,16 @@
 
   if (info->mPolicy == DelegatePolicy::ePersistDeniedCrossOrigin &&
       !mDocument->IsTopLevelContentDocument() &&
-      !mPrincipal->Subsumes(mTopLevelPrincipal)) {
+      IsCrossOriginContentToTop(mDocument)) {
     *aPermission = nsIPermissionManager::DENY_ACTION;
     return NS_OK;
   }
 
   nsIPrincipal* principal = mPrincipal;
-  if (mTopLevelPrincipal &&
-      (info->mPolicy == DelegatePolicy::eDelegateUseTopOrigin ||
+  if ((info->mPolicy == DelegatePolicy::eDelegateUseTopOrigin ||
        (info->mPolicy == DelegatePolicy::eDelegateUseFeaturePolicy &&
         StaticPrefs::dom_security_featurePolicy_enabled()))) {
-    principal = mTopLevelPrincipal;
+    // TODO: This will be changed in the next patch.
   }
 
   return (mPermissionManager->*testPermission)(principal, aType, aPermission);